src/controllers/midi/midisourceclock.cpp

```diff
@@ -0,0 +1,131 @@
+        mixxx::Duration earlyPulseTime;
+        if (m_iFilled < kRingBufferSize) {
+            earlyPulseTime = m_pulseRingBuffer[0];
+        } else {
+            // In a filled ring buffer, the earliest pulse is the next one that
+            // will get filled.
+            earlyPulseTime = m_pulseRingBuffer[m_iRingBufferPos];
+        }
+        m_dBpm = calcBpm(earlyPulseTime, lastPulseTime, m_iFilled);
+    }
+}
+
+// static
+double MidiSourceClock::calcBpm(const mixxx::Duration& early_pulse,
+                                const mixxx::Duration& late_pulse,
+                                int pulse_count) {
+    // Get the elapsed time between the latest pulse and the earliest pulse
+    // and divide by the number of pulses in the buffer to get bpm.  Midi
+    // clock information is by nature imprecise, and issues such as drift and
+    // inability to adapt to abrupt tempo changes are well known.  We can not
+    // expect to wring more precision out of an imprecise standard.
+
+    // If we have too few samples, we can't calculate a bpm, so return 0.0.
+    DEBUG_ASSERT_AND_HANDLE(pulse_count >= 2) {
+        qWarning() << "MidiSourceClock::calcBpm called with too few pulses";
+        return 0.0;
+    }
+
+    DEBUG_ASSERT_AND_HANDLE(late_pulse >= early_pulse) {
+        qWarning() << "MidiSourceClock asked to calculate beat fraction but "
+                   << "late_pulse < early_pulse:" << late_pulse << early_pulse;
+        return 0.0;
+    }
+
+    const mixxx::Duration elapsed = late_pulse - early_pulse;
+    const double elapsed_mins = elapsed.toDoubleSeconds() / 60.0;
+
+    // We subtract one since two time values denote a single span of time --
+    // so a filled value of 3 indicates 2 pulse periods, etc.
+    const double bpm = static_cast<double>(pulse_count - 1) / kPulsesPerQuarter
+                       / elapsed_mins;
+
+    if (bpm < kMinMidiBpm || bpm > kMaxMidiBpm) {
+        qWarning() << "MidiSourceClock bpm out of range, returning 0:" << bpm;
+        return 0;
+    }
+    return bpm;
+}
+
+// static
+double MidiSourceClock::beatFraction(const mixxx::Duration& last_beat,
+                                     const mixxx::Duration& now,
+                                     const double bpm) {
+    DEBUG_ASSERT_AND_HANDLE(now >= last_beat) {
+        qWarning() << "MidiSourceClock asked to calculate beat fraction but "
+                   << "now < last_beat:" << now << last_beat;
+        return 0.0;
+    }
+    // Get seconds per beat.
+    const double beat_length = 60.0 / bpm;
+    // seconds / secondsperbeat = fraction of beat.
+    const mixxx::Duration beat_duration = now - last_beat;
+    const double beat_percent = beat_duration.toDoubleSeconds() / beat_length;
+    // Ensure values are < 1.0.
+    return beat_percent - floor(beat_percent);
+}
+
+double MidiSourceClock::beatFraction() const {
+    const mixxx::Duration now = m_pClock->now();
+    return beatFraction(m_lastBeatTime, now, m_dBpm);
```

Comment:
I head a quick look into portmidi and it looks like it uses ftime() is a (deprecated) POSIX function on Linux. 
I assume that this is NTP corrected, so we should be aware that it can jump backwards. 

You can use PMEXPORT PtTimestamp Pt_Time(); from porttime.h, to adopt the Midi timestamp to the DAC time that drives Audio callback.